### ISSUE TITLE
issue-AI: Possible Bug in worklist implementation

### ISSUE THREAD

**Issue by mayank393** [CS639A-PAVT/BugTracker#26]
**Describe the bug**
The worklist algorithm throws an error when using a certain type of turtle program.

**To Reproduce**
- Provide the turtle program
```
if :a > 30 [
    if :a < 40 [
        forward :a
    ]
]
forward 10 <- remove this and the error goes away
```
- Add the following line in submissionAI.ForwardAnalysis.transferFunction
```
print("transfer", currBB.name)
```
- Add the following line in abstractInterpretation.py-> worklistAlgotithm
```
        for pred in predList: 
            label = cfg.get_edge_label(pred, currBB) #meant to find the location where to add the lines
            print("pred", pred) #add this 
            print("curr", currBB.name) #add this
```
**Logs**
```
Kachua v2.3
------------
if :a > 30 [
    if :a < 40 [
        forward :a
    ]
]
forward 10
========== IR ==========

0 (:a > 30)  [ 3 ]
1 (:a < 40)  [ 2 ]
2 forward :a  [ 1 ]
3 forward 10  [ 1 ]

transfer 0
pred 0
curr 1
transfer 1           <- log from transfer function
pred 0
curr 3
pred 1
curr 3
pred 2                <- notice that 2 is being considered as a predecessor, without calling transfer function on it
curr 3
Traceback (most recent call last):
  File "kachua.py", line 120, in <module>
    analyzeUsingAI(ir, args.progfl)
  File "<snip>/KachuaCore/../Submission/submissionAI.py", line 361, in analyzeUsingAI
    bbIn, bbOut = AI.worklistAlgorithm(cfg)
  File "<snip>/KachuaCore/abstractInterpretation.py", line 109, in worklistAlgorithm
    inlist.append(bbOut[pred.name][0])
IndexError: list index out of range
```
**Screenshots**
![possible-bug](https://user-images.githubusercontent.com/17690893/141761363-999457f4-b83f-412e-a558-b9e60fcad0f6.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04
 - Python3 version: 3.8.10

**Comment by mayank393** [comment]
Not sure if they are related but I am seeing a similar thing for loops.

```
repeat 3 [
    forward 1
]
========== IR ==========

0 :__rep_counter_1 = 3  [ 1 ]
1 (:__rep_counter_1 != 0)  [ 4 ]
2 forward 1  [ 1 ]
3 :__rep_counter_1 = (:__rep_counter_1 - 1)  [ 1 ]
4 False  [ -3 ]

transfer 0
pred 0
curr 1
pred 4             <-- using 4 without calling transfer on it. 
curr 1
Traceback (most recent call last):
  File "kachua.py", line 120, in <module>
    analyzeUsingAI(ir, args.progfl)
  File "<snip>/KachuaCore/../Submission/submissionAI.py", line 361, in analyzeUsingAI
    bbIn, bbOut = AI.worklistAlgorithm(cfg)
  File "<snip>/KachuaCore/abstractInterpretation.py", line 111, in worklistAlgorithm
    assert len(bbOut[pred.name]) > 1
AssertionError
```

**Comment by pkalita595** [comment]
Thank you for reporting the bug, fixed in Kachua-v5.3

**Comment by mayank393** [comment]
Thank you for fixing it so quick.
